### ISSUE TITLE
Question: disable nested handing of connection in class-based schema definition

### ISSUE THREAD

**Issue by yskkin** [rmosolgo/graphql-ruby#4188]
Open an issue for question as requested in https://github.com/rmosolgo/graphql-ruby/issues/1449#issuecomment-1239351523

I am trying to upgrade graphql to 1.8.
During migration of `GraphQL::ObjectType.define` to class-based one, I found issue about connection demonstrating in following test.

I expect that all cases pass, but `test_class_based` fails.
If `graphql_name 'ClassBasedConnection'` is changed to anything that does not end with Connection, all cases pass.

Can I upgrade LegacyConnection without changing graphql_name which breaks compatibility?
```ruby
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| 'https://github.com/#{repo}.git' }

  gem 'graphql', '~> 1.8.0'
  gem 'minitest'
end

require 'minitest/autorun'

LegacyConnection = GraphQL::ObjectType.define do
  name 'LegacyConnection'
  field :id, !types.String
end

class ClassBasedConnection < GraphQL::Schema::Object
  graphql_name 'ClassBasedConnection'
  field :id, String, null: false
end

Model = Struct.new(:id)

Schema = GraphQL::Schema.define do
  query GraphQL::ObjectType.define {
    name 'ApiQuery'

    connection :legacy, (GraphQL::Field.define do
      type LegacyConnection.connection_type
      argument :foo, types.String

      resolve(proc { [Model.new('a'), Model.new('b')] })
    end)

    connection :classBased, (GraphQL::Field.define do
      type ClassBasedConnection.connection_type
      argument :foo, types.String

      resolve(proc { [Model.new('a'), Model.new('b')] })
    end)
  }
end


class BugTest < Minitest::Test
  def test_legacy
    actual = Schema.execute(<<~QUERY, { variables: {} }).to_h
      query legacy {
        legacy {
          edges {
            node { id }
          }
        }
      }
    QUERY
    assert_equal({ 'data' => { 'legacy' => { 'edges' => [{ 'node' => { 'id' => 'a' } }, { 'node' => { 'id' => 'b' } }] } } }, actual)
  end

  def test_class_based
    actual = Schema.execute(<<~QUERY, { variables: {} }).to_h
      query classBased {
        classBased {
          edges {
            node { id }
          }
        }
      }
    QUERY
    assert_equal({ 'data' => { 'classBased' => { 'edges' => [{ 'node' => { 'id' => 'a' } }, { 'node' => { 'id' => 'b' } }] } } }, actual)
  end
end

```

**Comment by yskkin** [comment]
<details>
<summary>failure log</summary>

```
BugTest#test_class_based:
RuntimeError: No connection implementation to wrap Model (#<struct Model id="a">)
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/relay/base_connection.rb:36:in `connection_for_nodes'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/relay/connection_resolve.rb:37:in `build_connection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/relay/connection_resolve.rb:23:in `call'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/schema/member/instrumentation.rb:74:in `call'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/field.rb:248:in `resolve'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:297:in `call'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/schema/middleware_chain.rb:49:in `invoke_core'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/schema/middleware_chain.rb:38:in `invoke'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:108:in `block in resolve_field'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:57:in `block in trace'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:71:in `call_tracers'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:57:in `trace'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:107:in `resolve_field'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:71:in `block in resolve_selection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:64:in `each'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:64:in `resolve_selection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:260:in `resolve_value'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:189:in `continue_resolve_field'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:160:in `continue_or_wait'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:239:in `block in resolve_value'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:232:in `each'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:232:in `resolve_value'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:189:in `continue_resolve_field'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:160:in `continue_or_wait'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:133:in `resolve_field'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:71:in `block in resolve_selection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:64:in `each'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:64:in `resolve_selection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:260:in `resolve_value'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:189:in `continue_resolve_field'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:160:in `continue_or_wait'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:133:in `resolve_field'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:71:in `block in resolve_selection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:64:in `each'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:64:in `resolve_selection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:36:in `block in resolve_root_selection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:57:in `block in trace'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:71:in `call_tracers'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:57:in `trace'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/execute.rb:32:in `resolve_root_selection'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:108:in `begin_query'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:81:in `block in run_as_multiplex'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:80:in `map'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:80:in `run_as_multiplex'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:69:in `block (2 levels) in run_queries'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:175:in `block in instrument_and_analyze'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:29:in `block (2 levels) in apply_instrumenters'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:46:in `block (2 levels) in each_query_call_hooks'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:41:in `each_query_call_hooks'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:45:in `block in each_query_call_hooks'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:44:in `each_query_call_hooks'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:27:in `block in apply_instrumenters'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/instrumentation.rb:26:in `apply_instrumenters'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:167:in `instrument_and_analyze'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:68:in `block in run_queries'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:57:in `block in trace'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:71:in `call_tracers'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/tracing.rb:57:in `trace'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:58:in `run_queries'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/execution/multiplex.rb:48:in `run_all'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/schema.rb:347:in `block in multiplex'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/schema.rb:1132:in `with_definition_error_check'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/schema.rb:346:in `multiplex'
    /Users/yskkin/.rbenv/versions/2.7.6/lib/ruby/gems/2.7.0/gems/graphql-1.8.18/lib/graphql/schema.rb:323:in `execute'
    graphql-test.rb:64:in `test_class_based'
```

</details>

**Comment by rmosolgo** [comment]
Thanks for the detailed report! Recent graphql-ruby versions include `connection: false` to make sure that automatic connection handling doesn't apply to `node` fields: 

https://github.com/rmosolgo/graphql-ruby/blob/9cc79a95b66b9719815fd41801e227b20944e02e/lib/graphql/types/relay/edge_behaviors.rb#L29

It was fixed in #3223, first released in 1.11.7.  You could get the same behavior in 1.8.x by adding a monkey-patch like this one: 

```ruby
# Patch the GraphQL-Ruby built-in to make sure that node fields don't get treated like connections 
# (Remove this after 1.11.7+), see https://github.com/rmosolgo/graphql-ruby/pull/3223
class GraphQL::Types::Relay::BaseEdge
  def self.node_type(node_type = nil)
    if node_type
      @node_type = node_type
      wrapped_type_name = node_type.graphql_name
      # Set this to be named like the node type, but suffixed with `Edge`
      graphql_name("#{wrapped_type_name}Edge")
      # Add a default `node` field
      field :node, node_type, null: true, description: "The item at the end of the edge.", connection: false
    end
    @node_type
  end
end
```

If I add that after the `gemfile ...` block in your example, it works properly for me. Want to give it a try in your application?
